Marketplace

tjleing opened the pull request:
I accidentally merged the old marketplace branch into master instead of the other way around, so I'm opening a new pr for the rest of the changes.

### Summary
- These changes implement the marketplace functionality of donut.
- First, users can look at items and can filter listings by category and/or search queries.
- Also, users can add listings to the database, selecting item condition, price, title, and additional details.  If the item is a textbook, ISBN and author can also be selected.
- The user can pick the textbook that they are selling from a list, which specifies the title and author automatically, and if the textbook they are selling is not available, they can add it.
- Once a user creates a listing, they can also go in and edit it, changing any information they want to.

### Remaining to do
- [x] Enums for page names
- [x] I don't actually know if editing works, so that's kinda important
- [x] Let users edit via the item view page
- [x] Fail gracefully if item_id is invalid
- [x] Add tests
- [x] Using the logged-in user's id when submitting items
- [x] Make sure that when a user is editing, the user_id matches
- [x] Only display edit button if the user created the listing or is admin
- [x] Page for users to manage (edit or remove) their items
- [x] Also images exist (or should)

### Various smaller todos:
- [ ] get_table_list_data into _one and _all (or an argument)?
- [x] confirmation page (html and python)
- [ ] various refactoring
- [x] pass images to sell_2
- [x] pass images to sell_3
- [x] actually set things in the sql
- [x] pass images to view
- [x] handle adding and editing images
- [x] flash on inactive item
- [x] range(____|length) in Jinja to loop.index0
- [x] use directory search instead of the random endpoint things
- [ ] figure out table permissions
- [ ] even more tests?
- [x] cursor.lastrowid???
- [x] flexbox for categories instead of jank table
- [ ] use cat_order to sort categories (in top bar, item creation dropdown, etc) to ensure that selecting a category actually goes to the right place
- [ ] make sure that no change needs to be made due to permissions being different now

### CSS stuff to do after that's up (very exciting!)
- [x] tables probably shouldn't go all the way to the edges of the screen -- like 10% padding on each side?
  - [ ] have another look at that when CSS is up
- [ ] what's going on with the footer bar (esp. when creating a new item to sell)
- [ ] stylesheets instead of embedding CSS into HTML objects 😢

### Test Plan
- [x] Use the links to individual categories in the top box and make sure that items show up correctly.
- [x] Search for a few queries to see that querying works correctly (ex: 'name', 'gReAt', '').
- [x] Make sure that queries like 'alskghadfljasdfjk' that return no results are handled properly.
- [x] Make sure that items of category textbook and non-textbook can be added successfully and then appear when searching for them or just selecting the category or clicking on the 'here' link.
- [x] Make sure that new textbooks can be added to the database by using the form.
- [x] Make sure that you cannot continue to the next page of the form if any required data is missing, or if any data is malformed
- [x] Make sure that when not logged in, the "create an item" link redirects to a login page
- [x] Make sure that when logged in as a user different from the one who created an item, when attempting to edit their item, you aren't allowed to
- [x] Make sure the manage page displays all of the created listings
- [x] Make sure that for each of the manageable listings, you can edit, archive, unarchive or delete it
- [x] Make sure that archived items let you know they're archived when you edit or view them
- [x] Make sure that archived items don't show up in search results or category view pages
- [x] Make sure that you can upload images from your computer and from an imgur link
- [x] Make sure that you can see the images on the sell confirmation page and on the listing view page
- [x] Make sure that once an item has been added, hitting "edit this item" and then continuing through the form doesn't change the details of the item

tjleing committed:
An enum for keeping track of which page of the "create a new listing"
form a user is on.

tjleing committed:
Added updating the listing to helpers.py
so now editing works, except for the whole 'needing to be either the
user who made the listing or an admin' restriction thing

tjleing committed:
Added edit button to item view page

tjleing committed:
Catching some error messages and giving 404 instead

tjleing committed:
Merge remote-tracking branch 'origin/master' into marketplace

Conflicts:
	donut/modules/marketplace/routes.py

tjleing committed:
Okay so there was more than one merge conflict

tjleing committed:
Integer division can help sometimes

tjleing committed:
So yapf can make a build fail, who knew

tjleing committed:
Rename and a quick Makefile fix
might want to do a new branch for that but whatever
I think it's only affecting me so I can leave it here

tjleing committed:
More testing

tjleing committed:
Lint

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Fixed tests

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace
    donut.teardown_request(None)

tjleing committed:
So this was useless anyway

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Why is request.args different when you're testing

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace

tjleing committed:
Test fixes

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
More test changes

tjleing committed:
Make lint

dqu123 commented on Makefile:
```diff
@@ -13,7 +13,7 @@ lint:
 
 test: init-test-db
-	python -m pytest .
+	python3 -m pytest .
```
This shouldn't be necessary assuming your virtualenv is running.
Can you explain why you put this?

tjleing replied:
I guess my venv isn't working right for whatever reason.

tjleing replied:
I'll try messing around with the venv folder and see if I can remove that line.

tjleing replied:
Works now.

tjleing committed:
Fixed marketplace problem

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Bugfix and a style accident

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Fixes to validate_isbn, add_textbook, and tests

tjleing committed:
make lint

tjleing committed:
Some html fixes but a lot more tests

tjleing committed:
m a k e l i n t

tjleing committed:
More tests

tjleing committed:
Added login checking for selling items.

tjleing committed:
Fun time learning about flask session transactions

tjleing committed:
Only the person who created the listing or an admin can edit a listing

tjleing committed:
Only the person who created the listing or an admin can see the 'edit
this listing' button

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Bare-bones manage page which doesn't actually work yet

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace

tjleing committed:
LInt

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Manage page done!  All that's left to do is the archive and delete
pages, and then finally images.

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Moving code from routes to helpers to make it flow better, and some ' vs " consistency changes.

tjleing committed:
More " -> '

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Archiving and unarchiving done, only deleting left
but first all the sqlalchemy -> pymysql changes

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace

tjleing committed:
Rewrite get_table_list_data to use pymysql, and fixed a manage bug.

tjleing committed:
sqlalchemy -> pymysql, other bugfixes

tjleing committed:
Could this be the end of the sqlalchemy -> pymysql changes?

tjleing committed:
Well there was at least one more

tjleing committed:
And that's that!  Now I get to work on deleting items

tjleing committed:
Deleting items and various refactoring

tjleing committed:
So apparently I didn't know the difference between GET and POST
requests.  Anyway deleting is done now, archiving/unarchiving is
safer, and the code is cleaner

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Can upload images!!!

tjleing committed:
Lint

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Images now updated via SQL, commit comment changes made (thanks) (except
for the Jinja changes, which I'll do next)

tjleing committed:
Bugfixes, and more idiomatic Jinja (if that's even a thing)

tjleing committed:
Display images in confirmation page

tjleing committed:
Directory search integration; fixed image duplication bug

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Flash on view inactive item

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace

tjleing committed:
Major refactor part 1, I guess
Moved around a lot of functions in helpers
Began migration from get_table_list_data to table_fetch_one and
table_fetch_all
Fixed some things that weren't really bugs but were confusing
Extra documentation
Next up: one of the two reasons I did this refactor, generate_search_table()
(the other being sell())

tjleing committed:
Major refactor part 2.  Got through search,
finally used field_index_map on generate_search_table,
touched up a ton of comments, and moved even more functions
around.

tjleing committed:
Merge branch 'master' into marketplace

tjleing committed:
Fresh eyes, fresh refactors.  Thanks @calebsander for the suggestions.

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace

tjleing committed:
Use cursor.lastrowid instead of extra query

tjleing committed:
Lint :(

tjleing committed:
Merge branch 'master' into marketplace

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -1,30 +1,32 @@
 
 # taken from donut-legacy, which was apparently taken from a CS11
 # C++ assignment by dkong
 SKIP_WORDS = [
```
Convert this to a `set` for constant-time lookup

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -58,44 +54,259 @@ def render_with_top_marketplace_bar(template_url, **kwargs):
+    # in the html file.
+    cats2d = [[]] * num_rows
+    for cat_index in range(len(categories)):
+        cats2d[cat_index // num_cols].append(categories[cat_index])
```
Untested suggestion:
````python
cats2d = [categories[num_cols * row:num_cols * (row + 1)] for row in range(num_rows)]
````

tjleing replied:
No need anymore -- using flexbox to do it frontend now.  Getting rid of this old code feels good

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -58,44 +54,259 @@ def render_with_top_marketplace_bar(template_url, **kwargs):
+    """
+    current_user_id = get_user_id(flask.session['username'])
+    if current_user_id != get_user_id_of_item(
+            item_id) and not check_permission(Permissions.ADMIN):
```
Shouldn't this be an `or`? Or, `if not (current_user_id == ... and check_permission(...))`.

tjleing replied:
I think it's right -- if the person trying to make the change is neither the owner of the item nor an admin, don't make it.  I'll agree that it's a little confusing; do you think `if not (current_user_id == ... or check_permission(...))` would be better?

calebsander replied:
Yeah, I definitely misread this the first time I saw it. I figured it out later but guess I forgot to remove this comment. I think `if not (... or ...):` is a little more readable since it suggests you are doing something only in the exceptional case that the user is *not* allowed to do this. But totally your call.

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -58,44 +54,259 @@ def render_with_top_marketplace_bar(template_url, **kwargs):
+    field_index_map['archive'] = len(item) - 1
+
+    item.append('Delete')
+    field_index_map['delete'] = len(item) - 1
```
If these lines are switched, wouldn't need the `- 1`. (Same for 2 earlier instances)

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -58,44 +54,259 @@ def render_with_top_marketplace_bar(template_url, **kwargs):
+            '.manage_confirm', item=item_id, state='archive')
+    else:
+        links[field_index_map['archive']] = flask.url_for(
+            '.manage_confirm', item=item_id, state='unarchive')
```
Could use a ternary instead, since the left side is always the same

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -104,427 +315,176 @@ def generate_search_table(fields=None, attrs={}, query=""):
-                elif fields[field_index] == "item_timestamp":
-                    temp_res_row.append(data.strftime("%m/%d/%y"))
+        if 'item_id' in field_index_map:
+            item_id = int(item_listing[field_index_map['item_id']])
```
Won't `item_id in field_index_map` always be `True`? If this assignment doesn't occur, the `item_id` value from the previous loop will be reused, which seems wrong.

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+        result = cursor.fetchall()
+
+    # Return the list of lists.
+    result = list(map(lambda a: list(a.values()), result))
```
`[list(a.values()) for a in result]`. Also, don't think you can assume the order of the values in each row `dict`; I would use `[[row[field] for field in fields] for row in result]`.

tjleing replied:
Hmm, this only works if fields isn't `None`; above, if fields is `None` I use a `SELECT *`.  I think I'll use both of your suggestions -- `return [list(row.values()) for row in result]` if fields is `None` and `return [[row[field] for field in fields] for row in result]` otherwise.

calebsander replied:
Or, if `fields` was originally `None`, let `fields = list(result[0]) if result else []`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -104,427 +315,176 @@ def generate_search_table(fields=None, attrs={}, query=""):
-                elif fields[field_index] == "textbook_edition":
-                    temp_res_row.append(process_edition(data))
+            item_listing[field_index_map['user_id']] = get_name_from_user_id(
+                int(item_listing[field_index_map['user_id']]))
```
Also, are all these `int` casts necessary? I thought `pymysqldb` returned `int`s already for `int` columns.

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -104,427 +315,176 @@ def generate_search_table(fields=None, attrs={}, query=""):
+        # Can't merge, since the two columns aren't there.
+        return (datalist, fields, field_index_map)
 
     for row_index in range(len(datalist)):
```
`for row_index, row in enumerate(datalist):`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -104,427 +315,176 @@ def generate_search_table(fields=None, attrs={}, query=""):
+        if row[item_index] == '':
             row[item_index] = row[textbook_index]
         del row[textbook_index]
         datalist[row_index] = row
```
Don't think this is necessary; you never reassign to `row`, so it will already be the same value as `datalist[row_index]`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -538,24 +498,25 @@ def search_datalist(fields, datalist, query):
+        is_isbn_match = False
         if get_category_name_from_id(
-                listing[field_index_map["cat_id"]]) == "Textbooks":
+                listing[field_index_map['cat_id']]) == 'Textbooks':
```
Nit: might be more readable if you stored `listing[field_index_map['cat_id']]` in a variable so this condition wasn't split across 2 lines

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -538,24 +498,25 @@ def search_datalist(fields, datalist, query):
-                is_isbn_match = True
+                listing[field_index_map['item_title']])
 
         score = get_matches(query_tokens, item_tokens)
```
This should go in the `else` clause of `if is_isbn_match`, since it isn't necessary to compute if `is_isbn_match` is `True`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -611,6 +557,8 @@ def get_matches(l1, l2):
+    l1 = set(l1)
+    l2 = set(l2)
     if len(l1) < len(l2):
         return len([x for x in l1 if x in l2])
```
`len(l1 & l2)`

tjleing replied:
whoa

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -621,28 +569,21 @@ def tokenize_query(query):
-    query = query.split()
+    # Remove punctuation.
+    query = ' '.join(query)
+    punctuation_regex = r'[,.\-_!;:/\\]'
```
Maybe make this a global constant?

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -661,28 +602,50 @@ def validate_isbn(isbn):
+        total = 0
+        for i in range(10):
+            char = isbn[i].lower()
+            digit = 10  # x has value 10.
```
````python
for i, char in enumerate(isbn):
    digit = 10 if char.lower() == 'x' else int(char)
    # ...
````

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -661,28 +602,50 @@ def validate_isbn(isbn):
+        for i in range(13):
+            weight = 1
+            if i % 2 != 0:
+                weight = 3
```
`weight = 1 if i % 2 == 0 else 3`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -661,28 +602,50 @@ def validate_isbn(isbn):
+        if total % 10 == 0:
+            return True
+        else:
+            return False
```
`return total % 10 == 0`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
-                return str(edition) + "rd"
-            return str(edition) + "th"
+                return str(edition) + 'rd'
+            return str(edition) + 'th'
```
Rather than converting `edition` to `int` and then back to `str`, could just store the `int` in a separate variable

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+            return str(edition) + 'th'
         else:
             return str(edition)
     except ValueError:
```
No need to wrap all this code in a `try`-`except`; could just wrap the `edition = int(edition)` line

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+#################
+# TABLE QUERIES #
+#################
+def get_table_list_data(tables, fields=None, attrs={}):
```
Or default `fields` to `['*']` and you wouldn't need the special case

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    if fields == None:
+        s_select_columns = 'SELECT * '
+    else:
+        s_select_columns = 'SELECT ' + ', '.join(fields) + ' '
```
Move the trailing space to the start of `s_from` for consistency

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+
+    s_from = 'FROM ' + tables
+
+    s_where = ''
```
Can go in `else` clause of `if attrs`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+
+    # Return the list of lists.
+    result = list(map(lambda a: list(a.values()), result))
+    return result
```
Could just `return` rather than assigning to a temporary variable

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    return result
+
+
+def table_fetch_one(tables, fields=None, attrs={}):
```
This function almost completely duplicates `get_table_list_data()` -- would be better to write one function to handle their  common functionality

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    return result
+
+
+def table_fetch_all(tables, fields=None, attrs={}):
```
This function could easily be rewritten to call `get_table_list_data()`. I think the only difference is the special case for when `fields` is a singleton.

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+
+    # If only one field requested, unwrap [[_], [_], [_]] to [_, _, _]
+    if len(fields) == 1:
+        result = list(map(lambda a: a[0], result))
```
```suggestion
        result = [a[0] for a in result]
```

tjleing replied:
I would use the cool apply suggestion button, but I wanted to combine all these changes into one commit if possible.

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+        if parameter in skip_fields:
+            continue
+        if parameter in flask.request.form:
+            to_return.append([parameter, flask.request.form[parameter]])
```
Use a tuple here instead of a 2-element list?

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+        'marketplace_categories',
+        fields=['cat_title'],
+        attrs={'cat_id': category_id})
+    if len(cat_title) == 0:
```
`if cat_title` (and switch the `if` and `else` clauses)

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    if cat_title == 'Textbooks':
+        # regex to make sure the textbook edition is valid
+        if 'textbook_edition' in flask.request.form:
+            edition_regex = '^(|[0-9]+|international)$'
```
Could use `*` instead of `+` and remove the empty string case. Also, can this be made a global constant?

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+                        str(flask.request.form['textbook_edition']),
+                        re.IGNORECASE) == None:
+                errors.append(
+                    'Textbook edition is invalid. Try providing a number or \'International\'.'
```
Use double-quotes around the string to avoid having to escape the single-quotes

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    if not 'item_price' in flask.request.form or flask.request.form['item_price'] == '':
+        errors.append('Price cannot be left blank.')
+    else:
+        price_regex = '^([0-9]{1,4}\.[0-9]{0,2}|[0-9]{1,4}|\.[0-9]{1,2})$'
```
Can combine `[0-9]{1,4}\.[0-9]{0,2}|[0-9]{1,4}` into `[0-9]{1,4}(\.[0-9]{0,2})?`. Also, you should really use a regex string literal. It is not obvious to me that `'\.' == '\\.'` (in C and JavaScript, for instance, that is not the case).

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+            continue
+
+        if re.match(image_regex_1, image, re.IGNORECASE) == None:
+            if re.match(image_regex_2, image, re.IGNORECASE) != None:
```
Would be more readable if you negated this condition (and switched the clauses) or removed the `!= None`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    item_condition = stored['item_condition']
+    item_details = stored['item_details']
+    item_price = stored['item_price']
+    item_images = []
```
This statement is unnecessary

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+                user_id, cat_id, item_title, item_condition, item_details,
+                item_price
+            ])
+            item_id = cursor.lastrowid
```
These two clauses share a bit of duplicated code

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+    if item_id == -1:
+        return -1
+
+    s = 'INSERT INTO marketplace_images (item_id, img_link) VALUES (%s, %s);'
```
```suggestion
    s = 'INSERT INTO marketplace_images (item_id, img_link) VALUES (%s, %s)'
```
for consistency

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -693,25 +656,397 @@ def process_edition(edition):
+            continue
+        with flask.g.pymysql_db.cursor() as cursor:
+            cursor.execute(s, [item_id, image])
+    return item_id
```
This function duplicates most of the code in `create_new_listing()`. Would be nice to factor out their common functionality

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -726,21 +1061,72 @@ def add_textbook(title, author):
+    return result['user_id']
+
+
 def get_name_from_user_id(user_id):
```
Could instead use `core.helpers.get_name_and_email()`

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -768,13 +1155,15 @@ def get_textbook_info_from_textbook_id(textbook_id):
     if result == None:
         return None
-    return list(result)
+    return [result['textbook_title'], result['textbook_author']]
```
Tuples allow for better optimization:
```suggestion
    return (result['textbook_title'], result['textbook_author'])
```

calebsander commented on donut/modules/marketplace/helpers.py:
```diff
@@ -786,10 +1175,11 @@ def get_category_name_from_id(cat_id):
     if result == None:
         return None
-    return result[0]
+    return result['cat_title']
```
`return result and result['cat_title']`

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -86,76 +97,177 @@ def query():
+        stored_fields, {'item_id': item_id})
+
+    # make sure the item_id is a valid item, i.e. data is nonempty
+    if len(data) == 0:
```
```suggestion
    if not data:
```

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -86,76 +97,177 @@ def query():
+    # unwrap elements of 2d array
+    stored_images = list(map(lambda x: x[0], data))
+    # pad out to MAX_NUM_IMAGES
+    stored_images += [""] * (MAX_NUM_IMAGES - len(stored_images))
```
```suggestion
    stored_images += [''] * (MAX_NUM_IMAGES - len(stored_images))
```

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -86,76 +97,177 @@ def query():
+    data = helpers.get_table_list_data('marketplace_images', ['img_link'],
+                                       {'item_id': item_id})
+    # unwrap elements of 2d array
+    stored_images = list(map(lambda x: x[0], data))
```
```suggestion
    stored_images = [x[0] for x in data]
```

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -86,76 +97,177 @@ def query():
+        flask.session['next'] = flask.url_for('.sell')
+        return helpers.render_with_top_marketplace_bar('requires_login.html')
+
+    from donut.modules.marketplace.constants import Page
```
Conditional imports :(

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -166,42 +278,78 @@ def sell():
+                'marketplace_items NATURAL LEFT JOIN marketplace_textbooks',
+                stored_fields, {'item_id': item_id})
+
+            if len(data) == 0:
```
`if data:` and switch clauses

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -166,42 +278,78 @@ def sell():
+            # make sure the current user is the one who posted the item
+            current_user_id = get_user_id(flask.session['username'])
+            if stored['user_id'] != current_user_id and not check_permission(
+                    Permissions.ADMIN):
```
This check is used all over the place -- can it be factored into a separate function?

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -166,42 +278,78 @@ def sell():
+        # unwrap elements of 2d array
+        stored_images = list(map(lambda x: x[0], data))
+        # pad out to MAX_NUM_IMAGES
+        stored_images += [""] * (MAX_NUM_IMAGES - len(stored_images))
```
```suggestion
        stored_images += [''] * (MAX_NUM_IMAGES - len(stored_images))
```

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -166,42 +278,78 @@ def sell():
+        data = helpers.get_table_list_data('marketplace_images', ['img_link'],
+                                           {'item_id': item_id})
+        # unwrap elements of 2d array
+        stored_images = list(map(lambda x: x[0], data))
```
```suggestion
        stored_images = [x[0] for x in data]
```

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -86,76 +97,177 @@ def query():
+    if 'page' in flask.request.form:
         # but if we pass it in, get it
-        page = int(flask.request.form["page"])
+        page = Page.__members__[flask.request.form['page']]
```
```suggestion
        page = Page[flask.request.form['page']]
```

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -86,76 +97,177 @@ def query():
+        page = Page.__members__[flask.request.form['page']]
 
-    if not page in [1, 10, 2, 3, 4]:
+    if not page in Page:
```
Can't check this here; if `page` is not a valid page, `Page.__members__[flask.request.form['page']]` will throw a key error

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -210,38 +358,38 @@ def sell():
+                if textbook_title == '' or textbook_author == '':
                     flask.flash(
-                        "Textbook title and textbook author can't be blank.")
+                        'Textbook title and textbook author can\'t be blank.')
```
Double-quoted is nicer to avoid escaping the single quote

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -253,57 +401,61 @@ def sell():
-        if "textbook_id" not in flask.request.form or flask.request.form["textbook_id"] == "":
-            flask.flash("You need to select a textbook.")
+            if (page == Page.INFORMATION and prev_page == Page.CATEGORY) or (
+                    page == Page.CATEGORY and prev_page == Page.INFORMATION):
```
`if set([page, prev_page]) == set([Page.CATEGORY, page.INFORMATION]):`

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -253,57 +401,61 @@ def sell():
+                    page == Page.CATEGORY and prev_page == Page.INFORMATION):
+                page = Page.TEXTBOOK
+
+    if page in [Page.INFORMATION, Page.CONFIRMATION, Page.SUBMIT
```
`set` pls

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -313,20 +465,20 @@ def sell():
+                (stored['textbook_title'],
+                 stored['textbook_author']) = textbook_info[0]
             if field in flask.request.form:
                 stored[field] = flask.request.form[field]
```
`stored.update(flask.request.form)`

calebsander commented on donut/modules/marketplace/routes.py:
```diff
@@ -376,68 +531,49 @@ def sell():
-        stored["cat_title"] = cat_title
-        if (state == "new"):
+        stored['cat_title'] = cat_title
+        if (state == 'new'):
```
```suggestion
        if state == 'new':
```

calebsander commented on donut/modules/marketplace/templates/manage/confirm.html:
```diff
@@ -0,0 +1,43 @@
+            {% for item_field in item %}
+                {% set i = loop.index0 %}
+                <td>
+                    {% if links[i] %} <a href={{ links[i] }}> {% endif %}
```
```suggestion
                    {% if links[i] %} <a href="{{ links[i] }}"> {% endif %}
```

calebsander commented on donut/modules/marketplace/templates/manage/confirm.html:
```diff
@@ -0,0 +1,43 @@
+    </tbody>
+</table>
+
+<form action={{ url_for(".manage") }} method="post">
```
```suggestion
<form action="{{ url_for(".manage") }}" method="POST">
```

calebsander commented on donut/modules/marketplace/templates/manage/manage.html:
```diff
@@ -0,0 +1,87 @@
+                            {% set j = loop.index0 %}
+                            {% set link = activelinks[i][j] %}
+                            <td>
+                                {% if link %} <a href={{ link }}> {% endif %}
```
```suggestion
                                {% if link %} <a href="{{ link }}"> {% endif %}
```

calebsander commented on donut/modules/marketplace/templates/manage/manage.html:
```diff
@@ -0,0 +1,87 @@
+                            {% set j = loop.index0 %}
+                            {% set link = inactivelinks[i][j] %}
+                            <td>
+                                {% if link %} <a href={{ link }}> {% endif %}
```
```suggestion
                                {% if link %} <a href="{{ link }}"> {% endif %}
```

calebsander commented on donut/modules/marketplace/templates/requires_login.html:
```diff
@@ -0,0 +1,9 @@
+
+<p class="login-info">
+    This page requires you to login with your username and password.
+    Login <a href={{ url_for("auth.login") }}>here</a> and you'll be redirected.
```
```suggestion
    Login <a href="{{ url_for("auth.login") }}">here</a> and you'll be redirected.
```

calebsander commented on donut/modules/marketplace/templates/search.html:
```diff
@@ -19,13 +19,16 @@ <h4>
-                                {% if links[i][j] != 0 %} <a href={{ links[i][j] }}> {% endif %}
-                                {{ datalist[i][j] }}
-                                {% if links[i][j] != 0 %} </a> {% endif %}
+                                {% if link %} <a href={{ link }}> {% endif %}
```
```suggestion
                                {% if link %} <a href="{{ link }}"> {% endif %}
```

calebsander commented on donut/modules/marketplace/templates/sell/sell_1.html:
```diff
@@ -18,18 +18,20 @@ <h3>Modify an existing item</h3>
+                        {% set j = loop.index0 %}
                         <option value="{{ i*cats[0]|length+j+1 }}"
-                        {% if i*cats[0]|length+j+1 == category_id %} selected {% endif %} >
+                        {% if i*cat_row|length+j+1 == category_id %} selected {% endif %} >
```
Would be cleaner to set some variable to `i*cat_row|length+j+1`. Also, please put parens around `cat_row|length`.

calebsander commented on donut/modules/marketplace/templates/sell/sell_10.html:
```diff
@@ -55,7 +57,11 @@ <h3>Modify an existing item</h3>
+    <form action={{ url_for(".sell", state=state, item_id=request.args['item_id']) }} method='post'>
+    {% else %}
+    <form action={{ url_for(".sell", state=state) }} method='post'>
+    {% endif %}
```
These can be collapsed into a single `form` element. Use a ternary in the `action`.

calebsander commented on donut/modules/marketplace/templates/sell/sell_2.html:
```diff
@@ -77,10 +77,40 @@ <h3>Modify an existing item</h3>
+            var LOADING_PATH = "{{ url_for('static', filename='images/marketplace/loading.gif') }}";
+        </script>
+        <script src={{ url_for('static', filename='js/marketplace/overlay.js') }}></script>
+        <script src={{ url_for('static', filename='js/marketplace/sell_images.js') }}></script>
```
It's a good habit to wrap HTML attribute values in quotes

calebsander commented on donut/modules/marketplace/templates/topbar.html:
```diff
@@ -1,29 +1,31 @@
+                                {% for cat in row %}
+                                    {% set j = loop.index0 %}
+                                    <td style="width: {{ width }}%">
+                                        <a href={{ url_for(".category", cat=i*row|length+j+1) }}>{{ cat }}</a>
```
```suggestion
                                        <a href="{{ url_for(".category", cat=i*(row|length)+j+1) }}">{{ cat }}</a>
```

calebsander commented on donut/modules/marketplace/templates/topbar.html:
```diff
@@ -45,11 +47,13 @@ <h2>Marketplace</h2>
+                                {% for cat in row %}
+                                    {% set j = loop.index0 %}
+                                <option {% if i*row|length+j+1 == cat_id %} selected {% endif%}
+                                        value={{ i*row|length+j+1 }}>
```
Store `i*row|length+j+1` in a variable and use parens please

calebsander commented on donut/modules/marketplace/templates/view_item.html:
```diff
@@ -11,8 +11,11 @@ <h3>
-    <!-- TODO: when logins and privs are implemented, put "edit this item" here -->
+    {% if has_edit_privs %}
+    <h4>
+        <a href={{ url_for(".sell", state='edit', item_id=request.args["item_id"]) }}>Edit this listing</a>
```
```suggestion
        <a href="{{ url_for('.sell', state='edit', item_id=request.args['item_id']) }}">Edit this listing</a>
```

calebsander commented on donut/static/css/marketplace/overlay.css:
```diff
@@ -0,0 +1,53 @@
+    width: 100%;
+    height: 100%;
+    opacity: 0.7;
+    z-index: 9000;
```
ew

calebsander commented on donut/static/js/marketplace/overlay.js:
```diff
@@ -0,0 +1,73 @@
+    // Create the overlay. The overlayBackground and overlayWindow are child
+    // elements of the overlay div.
+    var overlay = document.createElement("div");
+    overlay.setAttribute("id", "overlay");
```
Can do stuff like `var overlay = $('<div>').attr('id', 'overlay')` in jQuery

calebsander commented on donut/static/js/marketplace/overlay.js:
```diff
@@ -0,0 +1,73 @@
+function centerOverlay()
+{
+    var width = $("#overlayWindow").width();
+    var height = $("#overlayWindow").height();
```
Store `$("#overlayWindow")` in a variable to avoid computing it twice

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+  showOverlay();
+
+  // Create the forms to upload images.
+  var uploadForm = "<div style=\"margin: 10px\">";
```
Single quotes here and later would allow you to not escape the HTML quotes

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+/*
+ * Display the form to upload an image.
+ */
+function showUploadForm(imageNum)
```
I don't know the code well enough to answer this, but are we SURE `imageNum` always comes from a secure source? Otherwise, there is some nasty XSS potential here since you are manually building HTML strings.

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+
+  // Set an error handler for ajax errors. This way, we let imgur handle
+  // invalid urls or files that the user may try to upload.
+  $(document).ajaxError(function() {
```
`$(document).ajaxError(handleImageUploadError)`

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+  uploadForm += "<tr>";
+  uploadForm += "<td>Upload from your computer:</td>";
+  uploadForm += "<td>Upload from the web:</td>";
+  uploadForm += "</tr><tr>";
```
Rather than `+=`ing each thing, just add a `+` at the end of the string on each line

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+    return false;
+  }
+
+  var imageFile = $("#image_file").prop("files")[0];
```
Store `$("#overlayWindow")` in a variable

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+  // the read is finished.
+  reader.onload = function(e) {
+    // Convert binary data to base 64 and then upload it.
+    uploadImage(btoa(e.target.result), imageNum, "base64");
```
```suggestion
    uploadImage(btoa(reader.result), imageNum, "base64");
```

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+    dataType: "json",
+    success: function(response) {
+      // Call the event handler with an additional argument.
+      console.log(response);
```
Remove after debugging is done

calebsander commented on donut/static/js/marketplace/sell_images.js:
```diff
@@ -0,0 +1,219 @@
+function handleUploadResponse(response, imageNum)
+{
+  // Check that the request was actually successful.
+  if (response.status != 200 || response.success != true)
```
```suggestion
  if (response.status != 200 || !response.success)
```

calebsander commented on sql/test_data.sql:
```diff
@@ -58,6 +58,13 @@ INSERT INTO member_options(user_id, option_id, option_type) VALUES
+    ('Textbooks', 2);
+
+INSERT INTO marketplace_items(cat_id, user_id, item_title, item_details, item_condition, item_price) VALUES
+    (1, 1, 'A table', "It\'s a pretty cool table, I guess", 'Good', 5.99);
```
Don't think `\` is necessary, but not sure

calebsander commented on tests/modules/marketplace/test_marketplace_helpers.py:
```diff
@@ -0,0 +1,107 @@
+    result = get_table_list_data(table, fields=fields, attrs=attrs)
+    assert len(result) == 1  # only one result
+    assert len(result[0]) == 2  # result has two fields
+    assert result[0] == [Decimal('5.99'), 'A table']
```
Combine into single assertion: `assert result == [[Decimal('5.99'), 'A table']]`

calebsander commented on tests/modules/marketplace/test_marketplace_helpers.py:
```diff
@@ -0,0 +1,107 @@
+    assert validate_isbn("9971-5-0210-0") == True
+    assert validate_isbn("0-8044-2957-X") == True
+    assert validate_isbn("0-9752298-0-X") == True
+    assert validate_isbn("978-3-16-148410-0") == True
```
`== True` is not needed

calebsander commented on tests/modules/marketplace/test_marketplace_helpers.py:
```diff
@@ -0,0 +1,107 @@
+    assert validate_isbn("9971-5-0210-X") == False
+    assert validate_isbn("0-8044-2957-1") == False
+    assert validate_isbn("0-9752298-0-2") == False
+    assert validate_isbn("978-3-16-148400-0") == False
```
Likewise, `== False` can be removed and `not` can be added to the front

calebsander commented on tests/modules/marketplace/test_marketplace_helpers.py:
```diff
@@ -0,0 +1,107 @@
+    assert process_edition(8.0) == "8th"
+    assert process_edition(9.0) == "9th"
+    assert process_edition(10.0) == "10th"
+    assert process_edition(2017.0) == "2017"
```
Are floats needed?

calebsander commented on tests/modules/marketplace/test_marketplace_routes.py:
```diff
@@ -0,0 +1,92 @@
+        # request (see http://flask.pocoo.org/docs/0.12/testing/#keeping-the-context-around)
+
+        rv = c.post(
+            flask.url_for('marketplace.sell'), data=dict(page='CATEGORY'))
```
Why not use a dict literal? Same below

tjleing committed:
Fix tests

tjleing committed:
Merge branch 'marketplace' of https://github.com/ASCIT/donut into marketplace

tjleing committed:
Only show active items on search results

tjleing committed:
Using flexbox instead of a weird 2d list to display categories in the
top bar

tjleing committed:
Lint

calebsander commented on donut/modules/marketplace/constants.py:
```diff
@@ -0,0 +1,9 @@
+from enum import Enum
+
+
+class Page(Enum):
```
Use `IntEnum` instead?
